types/underscore/index.d.ts

```diff
@@ -859,6 +860,11 @@ declare module _ {
          * @param lists The lists to zip.
          * @returns The zipped version of `lists`.
          **/
+        zip(): [];
+        zip<T, U, V>(...lists: [List<T>, List<U>, List<V>]): [T, U, V][];
+        zip<T, U>(...lists: [List<T>, List<U>]): [T, U][];
```

Comment:
Depending on how well TS is able to infer the types of lists, these two declarations might seriously break someone's code (because of the unequal length problem, as you mentioned yourself). Is there an easy way in which users can opt back out to the `List<any>[]` case?

Reply:
They could invoke `zip<any>(...)` instead if they wanted to opt out of stricter types.

I wonder though what code would be broken by this change? I can only think of situations where upgrading to this type definition would reveal a real bug. I guess this is a question around what you mean by "broken"?

For example, in the situation where lists of unequal length are provided, the return type should really be `[T?, U?][]`, which would force the user to check for undefined values before attempting to use the values for `T` and `U`. The risk of returning `[T, U][]` instead is that the user is not prompted to check for `undefined`, though they are welcome to do so (it is not a type error). But at least TS knows that the two zipped values should be of type `T` and `U` respectively.

In the current implementation, the return type is `any[][]`, so no type safety is given at all, let alone a reminder to check for `undefined` values.

Reply:
I tend to look at these things from the JS perspective. If I know that the arrays may be of unequal length, then I am aware that tuples at the end of the output array might contain `undefined` and I am probably writing code that takes this into account. With a TS typing of `any[][]`, this makes sense. If the TS typing changes to `[T, U][]` (where `T` and `U` exclude `undefined`), I suddenly find myself in disagreement with the TSC (and the TSC is technically wrong).

This might go by unnoticed (I guess this depends on strictness settings), but it could very well result in a warning or even an error down the line, which means that the TSC is going to tell me that my code is wrong when it is not.

That said, `zip<any>(...)` fits the bill of an opt out perfectly.

Reply:
That makes sense, thanks for elaborating!

How worried are you about future TS changes causing breakages vs present ones?

If it's just about the present, then I can't seem to find an example where working JS code results in TS compilation errors, even with strict turned on. My best shot is below. Within the `if` block, `a` adopts the `never` type, but TS doesn't throw an error even if you try to use the resulting value in an `undefined` context.

```
let c: undefined = undefined;
for (const [a, b] of _.zip([1, 2, 3], [4, 5])) {
    if (a === undefined) {
        c = a;
    }
}
```

If you're worried about the future, then I think TS will most likely not break this behaviour because of how they treat arrays. The following code also results in a `never` type, for example:

```
const array = [1, 2, 3];
const el = array[4];
if (el === undefined) {
  // el is never
}
```

I can also write a test to make sure that using zip with variable length lists does not break.

If I add such a test, combined with the reasoning above, do you think that's enough to address your concerns?

Reply:
I haven't looked at the latest changes yet, so what I write here might be outdated. With that out of the way:

> How worried are you about future TS changes causing breakages vs present ones?

Equally. 😉 

> 
> If it's just about the present, then I can't seem to find an example where working JS code results in TS compilation errors, even with strict turned on.

I can't think of anything off the top of my head, but I know that in my daily work, I run into these situations all the time. It's usually something subtle, indirect and unexpected and part of the problem is usually in DT packages. I don't blame the authors; JavaScript is dynamically typed, so it is impossible to do a perfect job with static type annotations, and the work is usually done for free. Still, the fact that part of the problem is in a DT package tends to make workarounds really fiddly and cumbersome. That, or I have to resort to `any`.

So, while I cannot always foresee how a subtly incorrect typing like this one will cause problems, I very much expect it to.

I take your word on the examples you wrote, but to be honest, I find them a bit simplistic; they don't seem to be representative of the problems I usually encounter.

> I can also write a test to make sure that using zip with variable length lists does not break.
> 
> If I add such a test, combined with the reasoning above, do you think that's enough to address your concerns?

Yes and no. I'm honestly a bit pessimistic that the changes won't cause problems, but with the easy escape that `zip<any>` provides, I accept the possibility. I have to write workarounds after minor and patch updates in DT packages all the time, so I consider it a fact of life.

Reply:
Thanks for the detailed reply, that makes a lot of sense, I'm sure it's a wild west out there. Fingers crossed this doesn't cause too much trouble, but as you say, there's an escape hatch.